paasta_tools/bounce_lib.py

```diff
@@ -339,19 +339,24 @@ def crossover_bounce(
     happy_new_tasks,
     old_app_live_happy_tasks,
     old_app_live_unhappy_tasks,
+    margin_factor=1.05,
 ):
     """Starts a new app if necessary; slowly kills old apps as instances of the new app become happy.
 
-    See the docstring for brutal_bounce() for parameters and return value.
+    :param margin_factor: the multiplication factor used to calculate the number of instances to be drained when minimum_health_capacity is reached.
+
+    See the docstring for brutal_bounce() for the rest of parameters and return value.
     """
 
+    assert margin_factor >= 1
+
     if not new_app_running:
         return {
             "create_app": True,
             "tasks_to_drain": set(),
         }
     else:
-        happy_count = len(happy_new_tasks)
+        happy_count = int(len(happy_new_tasks) * margin_factor)
         needed_count = max(new_config['instances'] - happy_count, 0)
```

Comment:
> When happy_count = 0 we gonna drain 5% of instances anyway.

I think that's okay, and desirable in some cases, such as if the cluster is very full.

> Moreover if new_app is broken, this will eventually drain the number of old_app instances to 0.

I don't think it would. `new_config['instances']` is based on the `instances` key in marathon.yaml or what the autoscaler decides, and isn't based on the number of running instances.